### PR TITLE
Add refresh script

### PR DESCRIPTION
**Pull request by JS-CodeGamer**
Add script to refresh cache on demand using a sitemap while limiting the api calls to a specified amount per minute

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,14 +1,16 @@
 #!/usr/bin/env node
-var prerender = require('./lib');
+var prerender = require("./lib");
 
 // set env variables
-process.env['ALLOWED_DOMAINS'] = "lessonpal.com,staging.lessonpal.com,dev.lessonpal.com,test.lessonpal.com,temp.lessonpal.com,localhost"
+process.env["ALLOWED_DOMAINS"] =
```
Comment (ravemz) [inline server.js]: Lets move this to .env file as well

File: server.js [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
 var server = prerender({
-	// pageDoneCheckInterval: 5000,
-	// waitAfterLastRequest: 5000,
-	// parseShadowDom: true,
-	// logRequests: true
+    // pageDoneCheckInterval: 5000,
+    // waitAfterLastRequest: 5000,
+    // parseShadowDom: true,
+    // logRequests: true
+    chromeLocation: "/usr/bin/google-chrome-stable",
```
Comment (ravemz) [inline server.js]: Move to .env

File: refresh-cache.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        await new Promise((resolve) => setTimeout(resolve, 1000 * 60));
+    }
+}
+
+async function refreshCacheForUrl(url) {
+    console.log("refreshing cache for", url);
+    return await axios({
+        url: prerender_url + "/render",
+        method: "post",
+        data: { url },
```
Comment (ravemz) [inline refresh-cache.js:60]: How are you taking care of invalidating the redis cache for this url? It has to be build in the prerenderRedisCache plugin.

**Review by ravemz — CHANGES_REQUESTED** [review]

File: refresh-cache.js (hunk 3 above)
Comment (JS-CodeGamer) [reply refresh-cache.js:60]: @ravemz as the server is resetting the same key it is like discarding the old key.
Found this while testing that the server ignores post requests and only fetches get request from the redis cache (as it should) so we can use that to refresh the entry in the database whithout causing any issues like the key does not exist even for a few seconds.

File: refresh-cache.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    );
+
+    for (let i = 0; i < sites.length / api_rate_limit; i++) {
+        await Promise.all(
+            sites
+                .slice(i * api_rate_limit, (i + 1) * api_rate_limit)
+                .map(refreshCacheForUrl)
+        );
+        // wait for 1min after api_rate_limit calls
+        await new Promise((resolve) => setTimeout(resolve, 1000 * 60));
```
Comment (ravemz) [inline refresh-cache.js:51]: Remove this await as we do not have a rate limit. And set api_rate_limit default to 20.

**Review by ravemz — CHANGES_REQUESTED** [review]

**Review by ravemz — APPROVED** [review]

File: refresh-cache.js (hunk 4 above)
Comment (JS-CodeGamer) [reply refresh-cache.js:51]: @ravemz I have removed the await but not sure if having the extra promise has any significance if you meant that the whole promise should be removed do tell me

**Pull request merged**